### PR TITLE
delete AggregateRoot

### PR DESCRIPTION
**Pull request by andrey-berenda**
move tests
add tests
small refactoring

### CODE REVIEW

File: ddd/domain_event_collection.py [hunk 1]
```diff
@@ -5,21 +5,48 @@
...
 
     def register(self, event: DomainEvent):
-        self._events.append(event)
-
-    def clear(self) -> None:
-        self._events.clear()
+        if not self._locked:
+            self._events.append(event)
+        else:
+            raise ValueError('You can not add events during the lock')
```
Comment (andrey-berenda) [inline ddd/domain_event_collection.py]: Здесь можно другой эксепшен, но кажется норма

File: ddd/domain_event_collection.py [hunk 2]
```diff
@@ -5,21 +5,48 @@
 from .domain_event import DomainEvent
 
 
+class DomainEventDescriptor:
+
```
Comment (pristupa) [inline ddd/domain_event_collection.py:13]: Could you add the same blank like to all other classes in the project please? I like it, so just to be consistent.

File: ddd/domain_event_collection.py (hunk 1 above)
Comment (pristupa) [reply ddd/domain_event_collection.py]: let's inverse the condition to remove "not" and the indentation for the main branch

File: ddd/domain_event_collection.py [hunk 3]
```diff
@@ -5,21 +5,54 @@
...
+        collection = instance.__dict__.get(self.field)
+
+        if collection is None:
+            instance.__dict__[self.field] = collection = DomainEventCollection()
+
+        return collection
+
+    def __set__(self, instance, value):
+        raise AttributeError("can't set attribute")
```
Comment (pristupa) [inline ddd/domain_event_collection.py]: Capitalize the first letter please

File: ddd/domain_event_collection.py [hunk 4]
```diff
@@ -5,21 +5,54 @@
...
+
 class DomainEventCollection(collections.abc.Collection):
+
     def __init__(self):
         self._events: List[DomainEvent] = []
+        self._locked = False
 
     def register(self, event: DomainEvent):
+        if self._locked:
+            raise Locked('You can not add events during the lock')
```
Comment (pristupa) [inline ddd/domain_event_collection.py]: `add -> register` for consistency

File: ddd/domain_event_handler.py [hunk 5]
```diff
@@ -38,14 +38,19 @@ def __set_name__(self, owner, name):
 
 def handle_domain_event(domain_event: DomainEvent) -> None:
     domain_event_class = type(domain_event)
-    if domain_event_class in _domain_event_handlers:
-        for handler_cls, handler in _domain_event_handlers[domain_event_class]:
-            handler_instance = _get_instance(handler_cls)
-            handler(handler_instance, domain_event)
+    if domain_event_class not in _domain_event_handlers:
+        raise ValueError(f'Not found domain event handler for class {domain_event_class}')
```
Comment (pristupa) [inline ddd/domain_event_handler.py]:
```suggestion
        raise ValueError(f'Domain event handler for class {domain_event_class} not found')
```
or
```suggestion
        raise ValueError(f'Unknown domain event {domain_event_class}. Please register at least one domain event handler')
```

File: tests/test_process_domain_events.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    aggregate1.domain_events.register(domain_event1)
+    aggregate2.domain_events.register(domain_event2)
+
+    with pytest.raises(MaximumRecursionException):
+        process_domain_events(aggregate1.domain_events, aggregate2.domain_events)
+
+
+@pytest.mark.usefixtures("instance_getter")
+def test_instance_set():
+    process_domain_events()
```
Comment (pristupa) [inline tests/test_process_domain_events.py]: newline

File: tests/test_process_domain_events.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+def test_try_set_domain_events_for_entity_with_domain_events():
+    aggregate = Aggregate(uuid.uuid4())
+    with pytest.raises(AttributeError):
+        aggregate.domain_events = []
+    assert isinstance(Aggregate.domain_events, DomainEventDescriptor)
+
+
+@pytest.mark.usefixtures("recursive_handler")
+def test_adding_domain_event_during_iteration_domain_events():
```
Comment (pristupa) [inline tests/test_process_domain_events.py:102]: how does it work?

File: tests/test_process_domain_events.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        @domain_event_handler
+        def empty_handle(self, domain_event: DomainEvent):
+            pass
+
+    yield Handler
+    delete_domain_event_handler(DomainEvent)
+
+
+class Aggregate(Entity):
+    domain_events = DomainEventDescriptor()
```
Comment (pristupa) [inline tests/test_process_domain_events.py]: How about not scaring developers by using a word "Descriptor"? What about just `domain_events = DomainEvents()`? Or do you think it makes the code less clear and straightforward?

File: tests/test_process_domain_events.py (hunk 7 above)
Comment (andrey-berenda) [reply tests/test_process_domain_events.py:102]: See recursive_handler
This handler add RecursiveDomainEvent to entity(second argument). So this handler try to register event for aggregate during the iterating

File: tests/test_process_domain_events.py (hunk 8 above)
Comment (andrey-berenda) [reply tests/test_process_domain_events.py]: I think DomainEvents is good
I'll rename

File: ddd/domain_event_collection.py [hunk 9]
```diff
@@ -5,21 +5,54 @@
...
+
 class DomainEventCollection(collections.abc.Collection):
+
     def __init__(self):
         self._events: List[DomainEvent] = []
+        self._locked = False
 
     def register(self, event: DomainEvent):
+        if self._locked:
+            raise Locked('You register not add events during the lock')
```
Comment (pristupa) [inline ddd/domain_event_collection.py]: oops :)

**Review by pristupa — CHANGES_REQUESTED** [review]

**Pull request closed without merging**